## Gather

In [1]:
import pandas as pd
import numpy as np
import requests
import json
import tweepy
import time
import  re


In [2]:
Tweet_Arch = pd.read_csv('DataSet/twitter-archive-enhanced.csv')
Tweet_Arch.head(2)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None  
1                  10    Tilly  None    None   None  None

In [3]:

url="https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv"
IMG = requests.get(url)

with open('image-predictions.tsv', 'wb') as file_IMG:
    file_IMG.write(IMG.content)

Imges_pred = pd.read_csv('image-predictions.tsv', sep = '\t')
Imges_pred.head(2)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True

In [4]:
#import tweepy

#consumer_key = ""
#consumer_secret = ""
#access_token = ""
#access_secret = ""

#auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
#auth.set_access_token(access_token, access_secret)

#api = tweepy.API(auth, parser=tweepy.parsers.JSONParser(),wait_on_rate_limit=True)

In [5]:
 
#json_list = [] 
#execption = []
 
#for tweet_id in Tweet_Arch['tweet_id']:
#    try:
#        line = api.get_status(tweet_id, tweet_mode = 'extended') 
#        favorite_count = line['favorite_count']
#        retweet_count = line['retweet_count']    
#        created_at = line['created_at']   
#        json_list.append({'tweet_id': int(tweet_id),
#                        'favorite_count': int(favorite_count),
#                        'retweet_count': int(retweet_count), 
#                        'created_at': pd.to_datetime(created_at)}) 
#    except Exception as e:  
#        print(str(e) + '->' + str(tweet_id))

In [6]:
#df_tweets = pd.DataFrame(json_list, columns = ['tweet_id', 'favorite_count', 'retweet_count' , 'date_time'])
#df_tweets.to_csv('tweet_json.txt', encoding = 'utf-8', index=False)

In [7]:
df_tweets = pd.read_csv('tweet_json.txt', encoding = 'utf-8')
df_tweets.head()

tweet_id  favorite_count  retweet_count  date_time
0  892420643555336193           37986           8302        NaN
1  892177421306343426           32611           6130        NaN
2  891815181378084864           24556           4058        NaN
3  891689557279858688           41324           8442        NaN
4  891327558926688256           39502           9147        NaN

In [8]:
df_tweets.shape

(2340, 4)

## Assess Data 

### 1- Twitter Archive

In [9]:
Tweet_Arch.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [10]:
Tweet_Arch.shape

(2356, 17)

In [11]:
Tweet_Arch.tail()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
2351  https://twitter.com/dog_rates/status/666049248...                 5   
2352  https://twitter.com/dog_rates/status/666044226...                 6   
2353  https://twitter.com/dog_rates/status/666033412...                 9   
2354  https://twitter.com/dog_rates/status/666029285...                 7   
2355  https://twitter.com/dog_rates/status/666020888...                 8   

      rating_denominator  name doggo floofer pupper puppo  
2351                  10  None  None    None   None  None  
2352                  10     a  None    None   None  None  
2353                  10     a  None    None   None  None  
2354                  10     a  None    None   None  None  
2355                  10  None  None    None   None  None

In [12]:
Tweet_Arch.sample(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1529  690248561355657216                    NaN                  NaN   
1220  714606013974974464                    NaN                  NaN   
677   789280767834746880                    NaN                  NaN   
2106  670468609693655041                    NaN                  NaN   
2207  668627278264475648                    NaN                  NaN   

                      timestamp  \
1529  2016-01-21 19:04:15 +0000   
1220  2016-03-29 00:12:05 +0000   
677   2016-10-21 01:42:53 +0000   
2106  2015-11-28 05:05:47 +0000   
2207  2015-11-23 03:09:00 +0000   

                                                 source  \
1529  <a href="http://twitter.com/download/iphone" r...   
1220  <a href="http://twitter.com/download/iphone" r...   
677   <a href="http://twitter.com/download/iphone" r...   
2106  <a href="http://twitter.com/download/iphone" r...   
2207  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
1529  This is Maxwell. That's his moped. He rents it...                  NaN   
1220  Here are two lil cuddly puppers. Both 12/10 wo...                  NaN   
677   RT @dog_rates: This is Meyer. He has to hold s...         7.507196e+17   
2106  This is Edd. He's a Czechoslovakian Googolplex...                  NaN   
2207  This is Timofy. He's a pilot for Southwest. It...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
1529                       NaN                        NaN   
1220                       NaN                        NaN   
677               4.196984e+09  2016-07-06 15:54:42 +0000   
2106                       NaN                        NaN   
2207                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
1529  https://twitter.com/dog_rates/status/690248561...                11   
1220  https://twitter.com/dog_rates/status/714606013...                12   
677   https://twitter.com/dog_rates/status/750719632...                12   
2106  https://twitter.com/dog_rates/status/670468609...                10   
2207  https://twitter.com/dog_rates/status/668627278...                 9   

      rating_denominator     name doggo floofer pupper puppo  
1529                  10  Maxwell  None    None   None  None  
1220                  10     None  None    None   None  None  
677                   10    Meyer  None    None   None  None  
2106                  10      Edd  None    None   None  None  
2207                  10   Timofy  None    None   None  None

In [13]:
Tweet_Arch.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [14]:
Tweet_Arch.dtypes

tweet_id                        int64
in_reply_to_status_id         float64
in_reply_to_user_id           float64
timestamp                      object
source                         object
text                           object
retweeted_status_id           float64
retweeted_status_user_id      float64
retweeted_status_timestamp     object
expanded_urls                  object
rating_numerator                int64
rating_denominator              int64
name                           object
doggo                          object
floofer                        object
pupper                         object
puppo                          object
dtype: object

In [15]:
Tweet_Arch['tweet_id'].value_counts()

749075273010798592    1
741099773336379392    1
798644042770751489    1
825120256414846976    1
769212283578875904    1
700462010979500032    1
780858289093574656    1
699775878809702401    1
880095782870896641    1
760521673607086080    1
776477788987613185    1
691820333922455552    1
715696743237730304    1
714606013974974464    1
760539183865880579    1
813157409116065792    1
676430933382295552    1
743510151680958465    1
837012587749474308    1
833722901757046785    1
818259473185828864    1
670704688707301377    1
667160273090932737    1
674394782723014656    1
672082170312290304    1
670093938074779648    1
759923798737051648    1
809920764300447744    1
805487436403003392    1
838085839343206401    1
                     ..
763956972077010945    1
870308999962521604    1
720775346191278080    1
785927819176054784    1
783347506784731136    1
775733305207554048    1
834209720923721728    1
825026590719483904    1
758405701903519748    1
668986018524233728    1
6909388994772213

In [16]:
Tweet_Arch['name'].value_counts()

None        745
a            55
Charlie      12
Cooper       11
Lucy         11
Oliver       11
Tucker       10
Lola         10
Penny        10
Bo            9
Winston       9
the           8
Sadie         8
Daisy         7
Buddy         7
Bailey        7
Toby          7
an            7
Stanley       6
Milo          6
Leo           6
Jack          6
Oscar         6
Scout         6
Dave          6
Jax           6
Koda          6
Rusty         6
Bella         6
Chester       5
           ... 
Bruiser       1
Mojo          1
Monkey        1
Josep         1
Petrick       1
Chadrick      1
Biden         1
Burt          1
Lorelei       1
Bilbo         1
Mingus        1
Tove          1
Tupawc        1
Beckham       1
Fido          1
Jay           1
Meera         1
Maks          1
Gunner        1
Einstein      1
such          1
Huck          1
Jazz          1
Tiger         1
Tanner        1
Pepper        1
Noosh         1
Tripp         1
Hanz          1
Grey          1
Name: name, Length: 957,

In [17]:
 
Tweet_Arch.loc[(Tweet_Arch['name'].str.islower())]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
22    887517139158093824                    NaN                  NaN   
56    881536004380872706                    NaN                  NaN   
118   869988702071779329                    NaN                  NaN   
169   859196978902773760                    NaN                  NaN   
193   855459453768019968                    NaN                  NaN   
335   832645525019123713                    NaN                  NaN   
369   828650029636317184                    NaN                  NaN   
542   806219024703037440                    NaN                  NaN   
649   792913359805018113                    NaN                  NaN   
682   788552643979468800                    NaN                  NaN   
759   778396591732486144                    NaN                  NaN   
773   776249906839351296                    NaN                  NaN   
801   772581559778025472                    NaN                  NaN   
819   770655142660169732                    NaN                  NaN   
822   770093767776997377                    NaN                  NaN   
852   765395769549590528                    NaN                  NaN   
924   755206590534418437                    NaN                  NaN   
988   748977405889503236                    NaN                  NaN   
992   748692773788876800                    NaN                  NaN   
993   748575535303884801                    NaN                  NaN   
1002  747885874273214464                    NaN                  NaN   
1004  747816857231626240                    NaN                  NaN   
1017  746872823977771008                    NaN                  NaN   
1025  746369468511756288                    NaN                  NaN   
1031  745422732645535745                    NaN                  NaN   
1040  744223424764059648                    NaN                  NaN   
1049  743222593470234624                    NaN                  NaN   
1063  741067306818797568                    NaN                  NaN   
1071  740214038584557568                    NaN                  NaN   
1095  736392552031657984                    NaN                  NaN   
...                  ...                    ...                  ...   
2191  668955713004314625                    NaN                  NaN   
2198  668815180734689280                    NaN                  NaN   
2204  668636665813057536                    NaN                  NaN   
2211  668614819948453888                    NaN                  NaN   
2212  668587383441514497                    NaN                  NaN   
2218  668507509523615744                    NaN                  NaN   
2222  668466899341221888                    NaN                  NaN   
2235  668171859951755264                    NaN                  NaN   
2249  667861340749471744                    NaN                  NaN   
2255  667773195014021121                    NaN                  NaN   
2264  667538891197542400                    NaN                  NaN   
2273  667470559035432960                    NaN                  NaN   
2287  667177989038297088                    NaN                  NaN   
2304  666983947667116034                    NaN                  NaN   
2311  666781792255496192                    NaN                  NaN   
2314  666701168228331520                    NaN                  NaN   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN         

In [18]:
Tweet_Arch['doggo'].value_counts()

None     2259
doggo      97
Name: doggo, dtype: int64

In [19]:
Tweet_Arch['floofer'].value_counts()

None       2346
floofer      10
Name: floofer, dtype: int64

In [20]:
Tweet_Arch['pupper'].value_counts()

None      2099
pupper     257
Name: pupper, dtype: int64

In [21]:
Tweet_Arch['puppo'].value_counts()

None     2326
puppo      30
Name: puppo, dtype: int64

In [22]:
Tweet_Arch.dtypes

tweet_id                        int64
in_reply_to_status_id         float64
in_reply_to_user_id           float64
timestamp                      object
source                         object
text                           object
retweeted_status_id           float64
retweeted_status_user_id      float64
retweeted_status_timestamp     object
expanded_urls                  object
rating_numerator                int64
rating_denominator              int64
name                           object
doggo                          object
floofer                        object
pupper                         object
puppo                          object
dtype: object

In [51]:
Tweet_Arch.loc[(Tweet_Arch['doggo'] == 'doggo') & (Tweet_Arch['floofer'] == 'floofer')]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
200  854010172552949760                    NaN                  NaN   

                     timestamp  \
200  2017-04-17 16:34:26 +0000   

                                                source  \
200  <a href="http://twitter.com/download/iphone" r...   

                                                  text  retweeted_status_id  \
200  At first I thought this was a shy doggo, but i...                  NaN   

     retweeted_status_user_id retweeted_status_timestamp  \
200                       NaN                        NaN   

                                         expanded_urls  rating_numerator  \
200  https://twitter.com/dog_rates/status/854010172...                11   

     rating_denominator  name  doggo  floofer pupper puppo  
200                  10  None  doggo  floofer   None  None

In [23]:
Tweet_Arch[Tweet_Arch.text.str.contains(r"(\d+\.\d*\/\d+)")][['text', 'rating_numerator']]


/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:1: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.
  """Entry point for launching an IPython kernel.


text  rating_numerator
45    This is Bella. She hopes her smile made you sm...                 5
340   RT @dog_rates: This is Logan, the Chow who liv...                75
695   This is Logan, the Chow who lived. He solemnly...                75
763   This is Sophie. She's a Jubilant Bush Pupper. ...                27
1689  I've been told there's a slight possibility he...                 5
1712  Here we have uncovered an entire battalion of ...                26

### 2- tweet image predictions

In [24]:
Imges_pred.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [25]:
Imges_pred.shape

(2075, 12)

In [26]:
Imges_pred.tail()

tweet_id                                          jpg_url  \
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2071  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2074  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num           p1   p1_conf  p1_dog                  p2   p2_conf  \
2070        2       basset  0.555712    True    English_springer  0.225770   
2071        1  paper_towel  0.170278   False  Labrador_retriever  0.168086   
2072        1    Chihuahua  0.716012    True            malamute  0.078253   
2073        1    Chihuahua  0.323581    True            Pekinese  0.090647   
2074        1       orange  0.097049   False               bagel  0.085851   

      p2_dog                           p3   p3_conf  p3_dog  
2070    True  German_short-haired_pointer  0.175219    True  
2071    True                      spatula  0.040836   False  
2072    True                       kelpie  0.031379    True  
2073    True                     papillon  0.068957    True  
2074   False                       banana  0.076110   False

In [27]:
Imges_pred.sample(5)

tweet_id                                          jpg_url  \
786   690360449368465409  https://pbs.twimg.com/media/CZSnKw8WwAAAN7q.jpg   
1321  756303284449767430  https://pbs.twimg.com/media/Cn7tyyZWYAAPlAY.jpg   
282   671134062904504320  https://pbs.twimg.com/media/CVBY3e7XIAAAE4Y.jpg   
831   693642232151285760  https://pbs.twimg.com/media/CaBP7i9W0AAJrIs.jpg   
1575  796031486298386433  https://pbs.twimg.com/media/CwwSaWJWIAASuoY.jpg   

      img_num                  p1   p1_conf  p1_dog                  p2  \
786         1                 pug  0.686933    True      French_bulldog   
1321        1    golden_retriever  0.981652    True      cocker_spaniel   
282         1            Shih-Tzu  0.180380    True    golden_retriever   
831         1  Scottish_deerhound  0.111893    True            bluetick   
1575        1    golden_retriever  0.893775    True  Labrador_retriever   

       p2_conf  p2_dog                           p3   p3_conf  p3_dog  
786   0.076359    True            Brabancon_griffon  0.035007    True  
1321  0.006790    True           Labrador_retriever  0.004325    True  
282   0.180194    True           Labrador_retriever  0.173656    True  
831   0.074302    True  German_short-haired_pointer  0.067000    True  
1575  0.070140    True                      doormat  0.008419   False

In [28]:
img_dup = Imges_pred["jpg_url"]
Imges_pred[img_dup.isin(img_dup[img_dup.duplicated()])]

tweet_id                                            jpg_url  \
85    667509364010450944    https://pbs.twimg.com/media/CUN4Or5UAAAa5K4.jpg   
224   670319130621435904    https://pbs.twimg.com/media/CU1zsMSUAAAS0qW.jpg   
241   670444955656130560    https://pbs.twimg.com/media/CU3mITUWIAAfyQS.jpg   
327   671896809300709376    https://pbs.twimg.com/media/CVMOlMiWwAA4Yxl.jpg   
382   673320132811366400    https://pbs.twimg.com/media/CVgdFjNWEAAxmbq.jpg   
432   674291837063053312    https://pbs.twimg.com/media/CVuQ2LeUsAAIe3s.jpg   
480   675354435921575936  https://pbs.twimg.com/ext_tw_video_thumb/67535...   
487   675501075957489664    https://pbs.twimg.com/media/CV_cnjHWUAADc-c.jpg   
587   679062614270468097    https://pbs.twimg.com/media/CWyD2HGUYAQ1Xa7.jpg   
591   679158373988876288    https://pbs.twimg.com/media/CWza7kpWcAAdYLc.jpg   
602   679828447187857408    https://pbs.twimg.com/media/CW88XN4WsAAlo8r.jpg   
713   685325112850124800    https://pbs.twimg.com/media/CYLDikFWEAAIy1y.jpg   
800   691416866452082688    https://pbs.twimg.com/media/CZhn-QAWwAASQan.jpg   
915   701214700881756160    https://pbs.twimg.com/media/Cbs3DOAXIAAp3Bd.jpg   
930   703041949650034688    https://pbs.twimg.com/media/CcG07BYW0AErrC9.jpg   
985   707610948723478529    https://pbs.twimg.com/media/CdHwZd0VIAA4792.jpg   
1033  711694788429553666  https://pbs.twimg.com/tweet_video_thumb/CeBym7...   
1045  712809025985978368    https://pbs.twimg.com/media/CeRoBaxWEAABi0X.jpg   
1118  725842289046749185    https://pbs.twimg.com/media/ChK1tdBWwAQ1flD.jpg   
1150  732005617171337216    https://pbs.twimg.com/media/CiibOMzUYAA9Mxz.jpg   
1155  733109485275860992    https://pbs.twimg.com/media/CiyHLocU4AI2pJu.jpg   
1188  739544079319588864    https://pbs.twimg.com/media/CkNjahBXAAQ2kWo.jpg   
1201  741067306818797568    https://pbs.twimg.com/media/CkjMx99UoAM2B1a.jpg   
1209  742423170473463808    https://pbs.twimg.com/media/Ck2d7tJWUAEPTL3.jpg   
1283  750429297815552001    https://pbs.twimg.com/media/CmoPdmHW8AAi8BI.jpg   
1297  752309394570878976  https://pbs.twimg.com/ext_tw_video_thumb/67535...   
1315  754874841593970688    https://pbs.twimg.com/media/CWza7kpWcAAdYLc.jpg   
1333  757729163776290825    https://pbs.twimg.com/media/CWyD2HGUYAQ1Xa7.jpg   
1345  759159934323924993    https://pbs.twimg.com/media/CU1zsMSUAAAS0qW.jpg   
1349  759566828574212096    https://pbs.twimg.com/media/CkNjahBXAAQ2kWo.jpg   
...                  ...                                                ...   
1699  816829038950027264    https://pbs.twimg.com/media/CvoBPWRWgAA4het.jpg   
1703  817181837579653120  https://pbs.twimg.com/ext_tw_video_thumb/81596...   
1705  817423860136083457  https://pbs.twimg.com/ext_tw_video_thumb/81742...   
1712  818588835076603904    https://pbs.twimg.com/media/Crwxb5yWgAAX5P_.jpg   
1715  819004803107983360    https://pbs.twimg.com/media/C12whDoVEAALRxa.jpg   
1716  819006400881917954    https://pbs.twimg.com/media/C12x-JTVIAAzdfl.jpg   
1717  819015331746349057    https://pbs.twimg.com/media/C12x-JTVIAAzdfl.jpg   
1718  819015337530290176    https://pbs.twimg.com/media/C12whDoVEAALRxa.jpg   
1727  820446719150292993    https://pbs.twimg.com/media/CxqsX-8XUAAEvjD.jpg   
1736  821813639212650496    https://pbs.twimg.com/media/CtVAvX-WIAAcGTf.jpg   
1738  822244816520155136    https://pbs.twimg.com/media/C2kzTGxWEAEOpPL.jpg   
1740  822489057087389700    https://pbs.twimg.com/media/C2oRbOuWEAAbVSl.jpg   
1742  822647212903690241    https://pbs.twimg.com/media/C2oRbOuWEAAbVSl.jpg   
1746  823269594223824897    https://pbs.twimg.com/media/C2kzTGxWEAEOpPL.jpg   
1755  824796380199809024    https://pbs.twimg.com/media/CwiuEJmW8AAZnit.jpg   
1767  826958653328592898    https://pbs.twimg.com/media/C3nygbBWQAAjwcW.jpg   
1785  829374341691346946    https://pbs.twimg.com/media/C4KHj-nWQAA3poV.jpg   
1789  829878982036299777    https://pbs.twimg.com/media/C3nygbBWQAAjwcW.jpg   
1791  830583320585068544    https://pbs.twimg.com/media/C4bTH6nWMA

In [34]:
Imges_pred.dtypes

tweet_id      int64
jpg_url      object
img_num       int64
p1           object
p1_conf     float64
p1_dog         bool
p2           object
p2_conf     float64
p2_dog         bool
p3           object
p3_conf     float64
p3_dog         bool
dtype: object

### 3- JSON tweets

In [29]:
df_tweets.head(5)

tweet_id  favorite_count  retweet_count  date_time
0  892420643555336193           37986           8302        NaN
1  892177421306343426           32611           6130        NaN
2  891815181378084864           24556           4058        NaN
3  891689557279858688           41324           8442        NaN
4  891327558926688256           39502           9147        NaN

In [30]:
df_tweets.tail(5)

tweet_id  favorite_count  retweet_count  date_time
2335  666049248165822465             106             42        NaN
2336  666044226329800704             293            139        NaN
2337  666033412701032449             123             43        NaN
2338  666029285002620928             126             47        NaN
2339  666020888022790149            2532            500        NaN

In [31]:
df_tweets.sample(5)

tweet_id  favorite_count  retweet_count  date_time
186   855818117272018944           27007           5592        NaN
1565  687127927494963200            5721           2437        NaN
1069  738402415918125056            3399            895        NaN
107   871102520638267392           20665           5460        NaN
417   821153421864615936             268             10        NaN

In [32]:
df_tweets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2340 entries, 0 to 2339
Data columns (total 4 columns):
tweet_id          2340 non-null int64
favorite_count    2340 non-null int64
retweet_count     2340 non-null int64
date_time         0 non-null float64
dtypes: float64(1), int64(3)
memory usage: 73.2 KB


In [33]:
df_tweets.dtypes

tweet_id            int64
favorite_count      int64
retweet_count       int64
date_time         float64
dtype: object

### Assessment Issues

### Quality
#####   1-Twitter Archive (Tweet_Arch data)
- comparing weet_Arch data with Imges_pred data 2356 compare to 2075 it same there is missing rows i belive this tweet without imeges.
- I notic in name column there is value like (a, an, the, such ,o,etc ..) is not realy names is seem like exracting issue
- in_reply_to_status_id , in_reply_to_user_id , retweeted_status_id , retweeted_status_user_id , retweeted_status_timestamp some of thim contain NAN value which is not usefull data.
- There is some rows have more than one dog.
- name varible have some values None it must be change to NAN.
- Missing data in column(expanded_urls)
- Erroneous datatypes tweet_id , timestamp,source,rating_denominator
 

#####   2- tweet image predictions (Imges_pred data) 
- some of jpg_url is duplicated 
- tweet_id convert to str 
- drop this columns p1', 'p1_conf', 'p1_dog', 'p2', 'p2_conf', 'p2_dog', 'p3', 'p3_conf', 'p3_dog'
#####   3- json API  (df_tweets data) 
- tweet_id convert to str 

### Tidiness
    
- Dog types column is seperated in three columns.
- we need to join and combined all datasets in one dataset 
 





## Clean

###### First we need to copy our data set to new data frame

In [80]:
Tweet_Arch_clean = Tweet_Arch.copy()
Imges_pred_clean = Imges_pred.copy()
df_tweets_clean = df_tweets.copy()

### Quality : Tweeter_Arch drop not needed column

##### Define
in_reply_to_status_id , in_reply_to_user_id , retweeted_status_id , retweeted_status_user_id , retweeted_status_timestamp some of thim contain NAN value which is not usefull data.

##### Code

In [81]:
unusefull_column = ['in_reply_to_status_id' , 'in_reply_to_user_id' ,'retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp']
Tweet_Arch_clean = Tweet_Arch_clean.drop(unusefull_column, axis=1)

##### Test

In [82]:
Tweet_Arch_clean.head()

tweet_id                  timestamp  \
0  892420643555336193  2017-08-01 16:23:56 +0000   
1  892177421306343426  2017-08-01 00:17:27 +0000   
2  891815181378084864  2017-07-31 00:18:03 +0000   
3  891689557279858688  2017-07-30 15:58:51 +0000   
4  891327558926688256  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   
1  This is Tilly. She's just checking pup on you....   
2  This is Archie. He is a rare Norwegian Pouncin...   
3  This is Darla. She commenced a snooze mid meal...   
4  This is Franklin. He would like you to stop ca...   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [83]:
Tweet_Arch_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 12 columns):
tweet_id              2356 non-null int64
timestamp             2356 non-null object
source                2356 non-null object
text                  2356 non-null object
expanded_urls         2297 non-null object
rating_numerator      2356 non-null int64
rating_denominator    2356 non-null int64
name                  2356 non-null object
doggo                 2356 non-null object
floofer               2356 non-null object
pupper                2356 non-null object
puppo                 2356 non-null object
dtypes: int64(3), object(9)
memory usage: 221.0+ KB


### Quality : Tweeter_Arch  Name Issue (a, an, the, such ,o,etc ..) is not real names is seem like extracting issue 

##### Define
I notice in name column there is value like (a, an, the, such ,o,etc ..) is not really names is seem like extracting issue and name varible have some values None it must be change to NAN.

##### Code

In [84]:
import re
Tweet_Arch_clean['txt_splt'] = Tweet_Arch_clean['text'].str.split()
all = []

def check_relly_name(df):           
    if 'named' in df['text'] and re.match(r'[A-Z].*', df['txt_splt'][(df['txt_splt'].index('named') + 1)]): 
            all.append(df['txt_splt'][(df['txt_splt'].index('named') + 1)])  
    elif df['text'].startswith('Here is a ') and re.match(r'[A-Z].*', df['txt_splt'][3]):
            all.append(df['txt_splt'][3].strip('.').strip(','))         
    elif df['text'].startswith('Here we have ') and re.match(r'[A-Z].*', df['txt_splt'][3]):
            all.append(df['txt_splt'][3].strip('.').strip(',')) 
    elif df['text'].startswith('This is ') and re.match(r'[A-Z].*', df['txt_splt'][2]):
            all.append(df['txt_splt'][2].strip('.').strip(',')) 
    elif df['text'].startswith('Say hello to ') and re.match(r'[A-Z].*', df['txt_splt'][3]):
            all.append(df['txt_splt'][3].strip('.').strip(',')) 
    elif df['text'].startswith('Meet ') and re.match(r'[A-Z].*', df['txt_splt'][1]):
            all.append(df['txt_splt'][1].strip('.').strip(',')) 
    else:
        all.append('Not_Provide') 
        
Tweet_Arch_clean.apply(check_relly_name, axis=1)

Tweet_Arch_clean['Extracted_name'] = all

In [85]:
Tweet_Arch_clean['name'].replace('O', "O'Malley", inplace = True)

In [86]:
Tweet_Arch_clean.loc[Tweet_Arch_clean.name.str.islower(),'name']= np.nan

In [87]:
Tweet_Arch_clean['name'].replace('None', np.nan, inplace = True) 

In [88]:
Tweet_Arch_clean.loc[Tweet_Arch_clean['Extracted_name'] == 'Not_Provide', 'Extracted_name'] = None 

##### Test

In [89]:
Tweet_Arch_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 14 columns):
tweet_id              2356 non-null int64
timestamp             2356 non-null object
source                2356 non-null object
text                  2356 non-null object
expanded_urls         2297 non-null object
rating_numerator      2356 non-null int64
rating_denominator    2356 non-null int64
name                  1502 non-null object
doggo                 2356 non-null object
floofer               2356 non-null object
pupper                2356 non-null object
puppo                 2356 non-null object
txt_splt              2356 non-null object
Extracted_name        1412 non-null object
dtypes: int64(3), object(11)
memory usage: 257.8+ KB


In [90]:
Tweet_Arch_clean.name.value_counts()

Charlie      12
Oliver       11
Lucy         11
Cooper       11
Lola         10
Penny        10
Tucker       10
Bo            9
Winston       9
Sadie         8
Daisy         7
Buddy         7
Toby          7
Bailey        7
Milo          6
Jack          6
Bella         6
Scout         6
Stanley       6
Koda          6
Dave          6
Rusty         6
Leo           6
Jax           6
Oscar         6
Oakley        5
Chester       5
Finn          5
Phil          5
Gus           5
             ..
Gin           1
Hubertson     1
Bruiser       1
Mojo          1
Monkey        1
Josep         1
Einstein      1
Hanz          1
Tripp         1
Noosh         1
Tito          1
Blakely       1
Burt          1
Bilbo         1
Tove          1
Tupawc        1
Beckham       1
Fido          1
Jay           1
Meera         1
Maks          1
Lorelei       1
Gunner        1
Fiji          1
Huck          1
Jazz          1
Tiger         1
Tanner        1
Pepper        1
Grey          1
Name: name, Length: 931,

In [91]:
Tweet_Arch_clean.Extracted_name.value_counts()

Charlie    11
Lucy       11
Cooper     10
Oliver     10
Penny       9
Tucker      9
Winston     8
Sadie       8
Lola        8
Daisy       7
Jax         6
Bella       6
Bailey      6
Bo          6
Oscar       6
Stanley     6
Koda        6
Toby        6
Rusty       5
Louis       5
Leo         5
Chester     5
Bentley     5
Milo        5
Scout       5
Buddy       5
Jeffrey     4
Clark       4
Derek       4
Gary        4
           ..
Chelsea     1
Moreton     1
Margo       1
Shadoe      1
Tanner      1
Tiger       1
Jazz        1
Hurley      1
Leonard     1
Fillup      1
Tito        1
Blakely     1
Burt        1
Bilbo       1
Tove        1
Carly       1
Tupawc      1
Beckham     1
Fido        1
Zeus.       1
Mona        1
Jay         1
Meera       1
Seamus      1
Maks        1
Lorelei     1
Gunner      1
Fiji        1
Huck        1
Grey        1
Name: Extracted_name, Length: 956, dtype: int64

### Quality : Tweeter_Arch  Erroneous datatypes 

##### Define
Erroneous datatypes tweet_id, timestamp, source,rating_numerator

##### Code

In [92]:
Tweet_Arch_clean['tweet_id'] = Tweet_Arch_clean['tweet_id'].astype(str)
Tweet_Arch_clean['source'] = Tweet_Arch_clean['source'].astype('category')  
Tweet_Arch_clean['timestamp'] = pd.to_datetime(Tweet_Arch_clean['timestamp'])
Tweet_Arch_clean['rating_numerator'] = Tweet_Arch_clean['rating_numerator'].astype('float')  

##### Test

In [93]:
Tweet_Arch_clean.dtypes

tweet_id                      object
timestamp             datetime64[ns]
source                      category
text                          object
expanded_urls                 object
rating_numerator             float64
rating_denominator             int64
name                          object
doggo                         object
floofer                       object
pupper                        object
puppo                         object
txt_splt                      object
Extracted_name                object
dtype: object

### Quality : Tweeter_Arch  Erroneous datatypes 

##### Define
Erroneous datatypes tweet_id, timestamp, source


##### Code

In [94]:
Tweet_Arch_clean['tweet_id'] = Tweet_Arch_clean['tweet_id'].astype(str)
Tweet_Arch_clean['source'] = Tweet_Arch_clean['source'].astype('category')  
Tweet_Arch_clean['timestamp'] = pd.to_datetime(Tweet_Arch_clean['timestamp'])

##### Test

In [95]:
Tweet_Arch_clean.dtypes

tweet_id                      object
timestamp             datetime64[ns]
source                      category
text                          object
expanded_urls                 object
rating_numerator             float64
rating_denominator             int64
name                          object
doggo                         object
floofer                       object
pupper                        object
puppo                         object
txt_splt                      object
Extracted_name                object
dtype: object

### Qulity : drop unused columns for Imges_pred_clean 

##### Define
there some column is seem not used in our analysis so I am gona drop it.

##### Code

In [96]:
Imges_pred_clean.drop(['p1', 'p1_conf', 'p1_dog', 'p2', 'p2_conf', 'p2_dog', 'p3', 'p3_conf', 'p3_dog',], axis=1, inplace=True)

##### Test

In [97]:
Imges_pred_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 3 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
dtypes: int64(2), object(1)
memory usage: 48.7+ KB


In [98]:
 Imges_pred_clean.shape

(2075, 3)

### Qulity :  Imges_pred_clean  drop duplicated images 

##### Define
there is some of jpg_url is duplicated need to remove it .

##### Code

In [99]:
 Imges_pred_clean.drop_duplicates(subset ="jpg_url",   keep = 'first', inplace = True)

##### Test

In [100]:
 Imges_pred_clean.shape

(2009, 3)

In [124]:
Imges_pred_clean[Imges_pred_clean['jpg_url'] == 'https://pbs.twimg.com/media/Cq9guJ5WgAADfpF.jpg']

tweet_id                                          jpg_url  \
1405  769940425801170949  https://pbs.twimg.com/media/Cq9guJ5WgAADfpF.jpg   

      img_num  
1405        1

### Qulity : df_tweets tweet_id datatype change

##### Define
we neeed to convert tweet_id of df_tweets to  str .

##### Code

In [101]:
df_tweets_clean['tweet_id'] = df_tweets_clean['tweet_id'].astype(str)

##### Test

In [102]:
df_tweets_clean.dtypes

tweet_id           object
favorite_count      int64
retweet_count       int64
date_time         float64
dtype: object

### Tidiness : Join Tweet_Arch_clean with Imges_pred_clean  some tweets without images


##### Define
we neeed to combined Tweet_Arch_clean with Imges_pred_clean some tweets without images but first we need convert tweet_id of Imges_pred_clean to str .

##### Code

In [103]:
Imges_pred_clean['tweet_id'] = Imges_pred_clean['tweet_id'].astype(str)

In [104]:
Tweet_Arch_clean = Tweet_Arch_clean.merge(Imges_pred_clean, on='tweet_id', how='inner')

##### Test

In [105]:
Imges_pred_clean.dtypes

tweet_id    object
jpg_url     object
img_num      int64
dtype: object

In [106]:
Tweet_Arch_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2009 entries, 0 to 2008
Data columns (total 16 columns):
tweet_id              2009 non-null object
timestamp             2009 non-null datetime64[ns]
source                2009 non-null category
text                  2009 non-null object
expanded_urls         2009 non-null object
rating_numerator      2009 non-null float64
rating_denominator    2009 non-null int64
name                  1350 non-null object
doggo                 2009 non-null object
floofer               2009 non-null object
pupper                2009 non-null object
puppo                 2009 non-null object
txt_splt              2009 non-null object
Extracted_name        1370 non-null object
jpg_url               2009 non-null object
img_num               2009 non-null int64
dtypes: category(1), datetime64[ns](1), float64(1), int64(2), object(11)
memory usage: 253.3+ KB


### Tidiness : Make doges column as one column

##### Define
After we join all dataset we need to merged  dog columns as one column by create new variable

##### Code

In [107]:
Tweet_Arch_clean['dog_type'] = Tweet_Arch_clean['text'].str.extract('(puppo|pupper|floofer|doggo)', expand=True)
dog_type_columns = ['doggo', 'floofer', 'pupper', 'puppo']
Tweet_Arch_clean = Tweet_Arch_clean.drop(dog_type_columns, axis=1)

##### Test

In [111]:
Tweet_Arch_clean['dog_type'].value_counts()

pupper     226
doggo       72
puppo       28
floofer      3
Name: dog_type, dtype: int64

In [112]:
Tweet_Arch_clean.head()

tweet_id           timestamp  \
0  892420643555336193 2017-08-01 16:23:56   
1  892177421306343426 2017-08-01 00:17:27   
2  891815181378084864 2017-07-31 00:18:03   
3  891689557279858688 2017-07-30 15:58:51   
4  891327558926688256 2017-07-29 16:00:24   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   
1  This is Tilly. She's just checking pup on you....   
2  This is Archie. He is a rare Norwegian Pouncin...   
3  This is Darla. She commenced a snooze mid meal...   
4  This is Franklin. He would like you to stop ca...   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...              13.0   
1  https://twitter.com/dog_rates/status/892177421...              13.0   
2  https://twitter.com/dog_rates/status/891815181...              12.0   
3  https://twitter.com/dog_rates/status/891689557...              13.0   
4  https://twitter.com/dog_rates/status/891327558...              12.0   

   rating_denominator      name  \
0                  10   Phineas   
1                  10     Tilly   
2                  10    Archie   
3                  10     Darla   
4                  10  Franklin   

                                            txt_splt Extracted_name  \
0  [This, is, Phineas., He's, a, mystical, boy., ...        Phineas   
1  [This, is, Tilly., She's, just, checking, pup,...          Tilly   
2  [This, is, Archie., He, is, a, rare, Norwegian...         Archie   
3  [This, is, Darla., She, commenced, a, snooze, ...          Darla   
4  [This, is, Franklin., He, would, like, you, to...       Franklin   

                                           jpg_url  img_num dog_type  
0  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg        1      NaN  
1  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg        1      NaN  
2  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg        1      NaN  
3  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg        1      NaN  
4  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg        2      NaN

In [113]:
Tweet_Arch_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2009 entries, 0 to 2008
Data columns (total 13 columns):
tweet_id              2009 non-null object
timestamp             2009 non-null datetime64[ns]
source                2009 non-null category
text                  2009 non-null object
expanded_urls         2009 non-null object
rating_numerator      2009 non-null float64
rating_denominator    2009 non-null int64
name                  1350 non-null object
txt_splt              2009 non-null object
Extracted_name        1370 non-null object
jpg_url               2009 non-null object
img_num               2009 non-null int64
dog_type              329 non-null object
dtypes: category(1), datetime64[ns](1), float64(1), int64(2), object(8)
memory usage: 206.2+ KB


### Quality : Tweets without images

#### comparing weet_Arch data with mges_pred data it same there is missing rows

##### Define
Delete the rows of tweets that not have image because it may be retweet or not tweet for rating purpose

##### Code

In [114]:
Tweet_Arch_clean = Tweet_Arch_clean.dropna(subset=['expanded_urls'])

##### Test

In [115]:
Tweet_Arch_clean.head()

tweet_id           timestamp  \
0  892420643555336193 2017-08-01 16:23:56   
1  892177421306343426 2017-08-01 00:17:27   
2  891815181378084864 2017-07-31 00:18:03   
3  891689557279858688 2017-07-30 15:58:51   
4  891327558926688256 2017-07-29 16:00:24   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   
1  This is Tilly. She's just checking pup on you....   
2  This is Archie. He is a rare Norwegian Pouncin...   
3  This is Darla. She commenced a snooze mid meal...   
4  This is Franklin. He would like you to stop ca...   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...              13.0   
1  https://twitter.com/dog_rates/status/892177421...              13.0   
2  https://twitter.com/dog_rates/status/891815181...              12.0   
3  https://twitter.com/dog_rates/status/891689557...              13.0   
4  https://twitter.com/dog_rates/status/891327558...              12.0   

   rating_denominator      name  \
0                  10   Phineas   
1                  10     Tilly   
2                  10    Archie   
3                  10     Darla   
4                  10  Franklin   

                                            txt_splt Extracted_name  \
0  [This, is, Phineas., He's, a, mystical, boy., ...        Phineas   
1  [This, is, Tilly., She's, just, checking, pup,...          Tilly   
2  [This, is, Archie., He, is, a, rare, Norwegian...         Archie   
3  [This, is, Darla., She, commenced, a, snooze, ...          Darla   
4  [This, is, Franklin., He, would, like, you, to...       Franklin   

                                           jpg_url  img_num dog_type  
0  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg        1      NaN  
1  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg        1      NaN  
2  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg        1      NaN  
3  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg        1      NaN  
4  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg        2      NaN

In [116]:
sum(Tweet_Arch_clean['expanded_urls'].isnull())

0

In [117]:
Tweet_Arch_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2009 entries, 0 to 2008
Data columns (total 13 columns):
tweet_id              2009 non-null object
timestamp             2009 non-null datetime64[ns]
source                2009 non-null category
text                  2009 non-null object
expanded_urls         2009 non-null object
rating_numerator      2009 non-null float64
rating_denominator    2009 non-null int64
name                  1350 non-null object
txt_splt              2009 non-null object
Extracted_name        1370 non-null object
jpg_url               2009 non-null object
img_num               2009 non-null int64
dog_type              329 non-null object
dtypes: category(1), datetime64[ns](1), float64(1), int64(2), object(8)
memory usage: 206.2+ KB


In [118]:
Tweet_Arch_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2009 entries, 0 to 2008
Data columns (total 13 columns):
tweet_id              2009 non-null object
timestamp             2009 non-null datetime64[ns]
source                2009 non-null category
text                  2009 non-null object
expanded_urls         2009 non-null object
rating_numerator      2009 non-null float64
rating_denominator    2009 non-null int64
name                  1350 non-null object
txt_splt              2009 non-null object
Extracted_name        1370 non-null object
jpg_url               2009 non-null object
img_num               2009 non-null int64
dog_type              329 non-null object
dtypes: category(1), datetime64[ns](1), float64(1), int64(2), object(8)
memory usage: 206.2+ KB


In [119]:
Tweet_Arch_clean.shape

(2009, 13)

#### Qulity : Name varible have some values None it must be change to NAN.

##### Define
 Name varible have some values None it must be change to NAN so we want to repace them to NAN

##### Code

In [120]:
Tweet_Arch_clean['name'] = Tweet_Arch_clean['name'].replace('None', np.NaN)
Tweet_Arch_clean['Extracted_name'] = Tweet_Arch_clean['name'].replace('None', np.NaN)
Tweet_Arch_clean.loc[Tweet_Arch_clean['dog_type'] == 'None', 'dog_type'] = None

##### Test

In [125]:
Tweet_Arch_clean.head()

tweet_id           timestamp  \
0  892420643555336193 2017-08-01 16:23:56   
1  892177421306343426 2017-08-01 00:17:27   
2  891815181378084864 2017-07-31 00:18:03   
3  891689557279858688 2017-07-30 15:58:51   
4  891327558926688256 2017-07-29 16:00:24   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   
1  This is Tilly. She's just checking pup on you....   
2  This is Archie. He is a rare Norwegian Pouncin...   
3  This is Darla. She commenced a snooze mid meal...   
4  This is Franklin. He would like you to stop ca...   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...              13.0   
1  https://twitter.com/dog_rates/status/892177421...              13.0   
2  https://twitter.com/dog_rates/status/891815181...              12.0   
3  https://twitter.com/dog_rates/status/891689557...              13.0   
4  https://twitter.com/dog_rates/status/891327558...              12.0   

   rating_denominator      name  \
0                  10   Phineas   
1                  10     Tilly   
2                  10    Archie   
3                  10     Darla   
4                  10  Franklin   

                                            txt_splt Extracted_name  \
0  [This, is, Phineas., He's, a, mystical, boy., ...        Phineas   
1  [This, is, Tilly., She's, just, checking, pup,...          Tilly   
2  [This, is, Archie., He, is, a, rare, Norwegian...         Archie   
3  [This, is, Darla., She, commenced, a, snooze, ...          Darla   
4  [This, is, Franklin., He, would, like, you, to...       Franklin   

                                           jpg_url  img_num dog_type  
0  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg        1      NaN  
1  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg        1      NaN  
2  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg        1      NaN  
3  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg        1      NaN  
4  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg        2      NaN

### Tidiness : Join Tweet_Arch_clean with df_tweets_clean 


##### Define
Join Tweet_Arch_clean with df_tweets_clean  to store it in one data set and do some analysis 

##### Code

In [127]:
Tweet_Arch_clean = pd.merge(left=Tweet_Arch_clean, right=df_tweets_clean, left_on='tweet_id', right_on='tweet_id', how='inner')


##### Test

In [130]:
Tweet_Arch_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2006 entries, 0 to 2005
Data columns (total 15 columns):
tweet_id              2006 non-null object
timestamp             2006 non-null datetime64[ns]
source                2006 non-null category
text                  2006 non-null object
expanded_urls         2006 non-null object
rating_numerator      2006 non-null float64
rating_denominator    2006 non-null int64
name                  1348 non-null object
txt_splt              2006 non-null object
Extracted_name        1348 non-null object
jpg_url               2006 non-null object
img_num               2006 non-null int64
dog_type              329 non-null object
favorite_count        2006 non-null int64
retweet_count         2006 non-null int64
dtypes: category(1), datetime64[ns](1), float64(1), int64(4), object(8)
memory usage: 237.2+ KB


### Save DataSet As csv file

In [131]:
Tweet_Arch_clean.to_csv('twitter_archive_master.csv')